pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -78,6 +82,10 @@ rec {
 
     , ... } @ attrs:
 
+    # Check that the name is consistent with pname and version:
+    assert lib.lists.all (name: builtins.hasAttr name attrs) ["name" "pname" "version"]
+      -> lib.strings.hasSuffix "${attrs.pname}-${attrs.version}" attrs.name;
```

Comment:
Are you sure this is the way to go?

I think not allowing both `pname` & `version` and `name` would lead to less confusion.

Also, use `lib.assertMsg` for this to add an error message.

Reply:
> I think not allowing both `pname` & `version` and `name` would lead to less confusion.

I (mostly) agree. However, that would require tree-wide changes to packages already using the following pattern:

```nix
mkDerivation rec {
  pname = "foo";
  version = "bar";
  name = "${pname}-${version}"

  # ...
}
``` 

Of which there are quite a few.

The follow-up to this PR is of course a series of PR's eliminating this pattern, but I don't think it should all be done in one PR (I tried to do that in the previous PR, and it went stale rather quickly, so I'd like to keep this PR relatively small).

After this all is fixed, I'd say we shouldn't even allow `pname` and `name`. 

> Also, use `lib.assertMsg` for this to add an error message.

Ah, yes, of course, fixing that!